### PR TITLE
Consensus Reactor communication specification

### LINKED ISSUE #9851: Formal specification of the Gossip layer.
This issues tracks the efforts to document and specify the interaction between the consensus state machine (State) and the communication layer (Gossip) inside the Consensus reactor.
This task will serve as input to #9880 

The object of this effort is v0.37.x.

- [x] #9930
- [x] #9931
- [x] #9932

DoD:
- The current implementation is well understood
    - all components of the current implementation are documented
    - the interaction between Gossip and State is documented.
- There exists a formal specification of the State/Gossip/P2P interactions and mismatches between requirements and provisions are identified.

### PR DESCRIPTION
**Pull request by lasarojc**
Closes #9851 

---

#### PR checklist

- [ ] Tests written/updated, or no tests needed
- [ ] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [ ] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

File: spec/consensus/reactor/reactor.md [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+=====================================================================
+                            Network Stack
+```
+
+This document focuses on the interactions between the Consensus Reactor and the P2P layer.
+The Consensus reactor is, itself, divided into two layers, State and Communication. 
+
+The **State** layer keeps the state and transition functions described in the [The latest gossip on BFT consensus](https://arxiv.org/abs/1807.0493).
+
+The **Communication** layer (AKA, Gossip) handles messages received through the P2P layer, triggering updates of peer information and, when conditions are met, of the consensus state.
```
Comment (josef-widder) [inline spec/consensus/reactor/reactor.md:28]: It is a bit unclear where things live. It seems the "peer information" is within communication, while the "consensus state" is in the state layer, right?

File: spec/consensus/reactor/reactor.md [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+The communication layer also reacts to events in the consensus state and, periodically, propagates information on the node's state to its neighbors.
```
Comment (josef-widder) [inline spec/consensus/reactor/reactor.md:29]: "events in the consensus state": wording is a bit unclear

File: spec/consensus/reactor/reactor.md [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+Because Gossip Communication requires even messages sent before GST to be reliably delivered between correct processes ($t$ may happen before GST in (i)) and because GST could take arbitrarily long to arrive, in practice, implementing this property would require an unbounded message buffer.
+
+However, while Gossip Communication is a sufficient condition for Tendermint BFT to terminate, it is not strictly necessary that all messages from correct processes reach all other correct processes.
+What is required is for either the message to be delivered or that, eventually, some newer message, with information that supersedes that in the first message, to be timely delivered.
+In Tendermint BFT, this property is seen, for example, when nodes ignore proposal messages from prior rounds.
+
+> **Supersession**: We say that a message $lhs$ supersedes a message $rhs$ if after receiving $lhs$ a process would make at least as much progress as it would by receiving $rhs$ and we note is as follows: $\text{lhs} \overset{s}{>} \text{rhs}$.
```
Comment (josef-widder) [inline spec/consensus/reactor/reactor.md:82]: In the past we looked a bit into communication closure w.r.t. consensus. Roughly, the lexicographical order over the tuples (height, round, step) defines a notion of logical time, and when I am in a certain height, round and step, I don't care about messages from "the past". Tendermint consensus is mostly communication-closed in that we don't care about messages from the past. An exception is line 28 in the arXiv paper where we accept prevote messages from previous rounds vr for the same height.

I guess a precise constructive definition of "supersession" can be done along these lines.

File: spec/consensus/reactor/reactor.md [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+A message supersession operator:
+
+`SSS(lhs,rhs)` (`lhs` supersedes `rhs`) is true if and only if $\text{lhs} \overset{s}{>} \text{rhs}$
+
+### Requires from the Communication Layer
+
+Messages received by a node either change the node's State or are discarded. 
+If they change the State, even if by just getting added to a set of received messages, then the original message need not be forwarded. 
+Instead, a new message, with the same contents, is generated from the State of the node. Why does this matter? Because it let's us think in terms of messages being sent to neighbors only, not forwarded on the overlay.
```
Comment (josef-widder) [inline spec/consensus/reactor/reactor.md:111]: I agree that this allows us to think about the neighbors only. At the same time it introduces a level of disconnection from the properties in the arXiv paper

File: spec/consensus/reactor/reactor.md [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+```tla+
+GossipCast(p,m): Msgs[p]' = Msgs[p] \cup {m} 
+```
+
+All outbound messages will be delivered to neighbors at the time the message was sent, or the neighbor disconnects, or a superseding messages makes sending the message useless.
+
+```tla+
+GossipCastDelivery(p,q,m) == 
+    q \in Ne[p] /\ m \in Msgs[p] ~> \/ m \in DMsgs[q]
+                                    \/ m \notin Msgs[p] /\ \E mn \in Msgs[p]: SuperS(mn,m)
```
Comment (josef-widder) [inline spec/consensus/reactor/reactor.md:146]:
```suggestion
                                    \/ \E mn \in Msgs[p]: SuperS(mn,m)
```
It seems to me that logically the "\notin" part is not needed

File: spec/consensus/reactor/reactor.md [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    - Enqueue and forget. 
+    - Disconnection and reconnection may cause drops. 
+    - Enqueuing may block for a while in `Send`, but not on `TrySend`
+
+
+
+#### Non-requirements
+- Non-duplication
+    - Gossip itself will duplicate messages
+    - Idempotency must be implemented by the application
```
Comment (josef-widder) [inline spec/consensus/reactor/reactor.md:257]: I guess here by application you mean consensus?

File: spec/consensus/reactor/reactor.md (hunk 1 above)
Comment (lasarojc) [reply spec/consensus/reactor/reactor.md:28]: Correct. Rephrased to clarify.

File: spec/consensus/reactor/reactor.md (hunk 2 above)
Comment (lasarojc) [reply spec/consensus/reactor/reactor.md:29]: I was influenced by the use of pub-sub to send messages. Abstracted and rephrased.

File: spec/consensus/reactor/reactor.md (hunk 5 above)
Comment (lasarojc) [reply spec/consensus/reactor/reactor.md:146]: True. We don't need to lose the message, just not deliver it.

File: spec/consensus/reactor/reactor.md (hunk 6 above)
Comment (lasarojc) [reply spec/consensus/reactor/reactor.md:257]: Correct. Rephrased for clarity.

**Comment by josef-widder** [comment]
This seems to be the first Tendermint spec that should include the new specification language (not sure what the final name is). I tag @shonfeder as he is looking for this kind of information.

**Pull request closed without merging**